README.md
# Redux-Response-Middleware
Single Redux Middleware with Request/Success/Failure Pattern

## Installation

```
npm i redux-response-middleware --save
```

## Use

Add the middleware.

```javascript
import { combineReducers, createStore, applyMiddleware } from 'redux';

import users from '@redux/users/reducer';
import posts from '@redux/posts/reducer';
import responseMiddleware from '@redux/helpers/responseMiddleware';

const reducers = combineReducers({ users, posts });

export default createStore(reducers, applyMiddleware(responseMiddleware()));

```

If the action has a `service` property you will receive corresponding props: data, error, loading, if these properties doesn't exist the action going to work as usually .

```javascript
import { GET_POSTS, ADD_POST } from '@constants/actionNames';
import postServices from '@services/postServices';

const getPostsTarget = 'posts';
const addPostsTarget = 'post';

const success = (response) => {
  console.log('success:', response);
};

const failure = (error) => {
  console.log('error', error);
};

const postActions = {
  getPosts: (success, failure) => ({
    type: GET_POSTS,
    target: getPostsTarget,
    service: postServices.getPosts,
    response: resp => resp.data,
    error: error => error.data,
    success,
    failure,
  }),
  addPost: (values, success, failure) => ({
    type: ADD_POST,
    target: addPostsTarget,
    service: postServices.addPost(values),
    response: resp => resp.data,
    error: error => error.data,
    success,
    failure,
  }),
};

export default postActions;
```

Option | Type | Description
------ | ---- | ------- 
type | String | You should send a type name to handle from your reducer.
service | Promise | You should send a promise, for example `axios.get('posts')`.
target | String | This is important to get good prop names, for example if `target: 'post'` you get props: postData, postError, postLoading.
initialState | * | The initial value ([], {}, '', etc)
response | Func (optional) | is a callback to handle the response route, for example if you have `data: { data: { title } } ` you could  use `resp => resp.data.data`, so that you can get in your redux state `postData: { title }` instead of `postData: data: { data: { title } }`.
error | Func (options) | Same as response option, but to handle errors.
success | Func (response) | The callback when the response is success.
failure | Func (error) | The callback when you get error


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
  
## [1.3.0] - 2021-07-25
  
Here we would have the update steps for 1.2.4 for people to follow.
 
### Added
-
 
### Changed
  
- [#f187f05](https://github.com/Christianmo/Redux-Response-Middleware/commit/f187f0589a25999575d21473ba2f4da084f16f60)
Now using Typescript
 
### Fixed
-
 


src/index.ts
import { Middleware } from 'redux';

function responseMiddleware() {
  const service = "service";
  const success = "success";
  const failure = "failure";
  const response = "response";
  const initialState = "initialState";

  const middleware: Middleware = () => next => action => {
    const hasProperty = (property: any) =>
      Object.prototype.hasOwnProperty.call(action, property);

    if (hasProperty(service)) {
      if (typeof action[service] === "object") {
        const data = `${action.target}Data`;
        const error = `${action.target}Error`;
        const loading = `${action.target}Loading`;

        const initialData = hasProperty(initialState)
          ? action[initialState]
          : false;

        const payload = {
          [data]: initialData,
          [error]: false,
          [loading]: false
        };

        next({
          ...action,
          payload: { ...payload, [loading]: true }
        });

        action[service]
          .then((resp: any) => {
            if (resp.status < 200 || resp.status > 299) {
              const error: Error = new Error();
              throw error;
            }

            next({
              ...action,
              payload: {
                ...payload,
                [data]: action.response ? action[response](resp) : resp
              }
            });

            if (hasProperty(success)) action[success](resp);
          })
          .catch((err: Error) => {
            next({
              ...action,
              payload: {
                ...payload,
                [error]: action.error ? action.error(err) : err
              }
            });

            if (hasProperty(failure)) action[failure](err);
          });
      } else {
        throw new Error("action.service should be a promise");
      }
    } else {
      next(action);
    }
  };

  return middleware;
}

export default responseMiddleware;


lib/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
function responseMiddleware() {
    var service = "service";
    var success = "success";
    var failure = "failure";
    var response = "response";
    var initialState = "initialState";
    return function () { return function (next) { return function (action) {
        var _a, _b;
        var hasProperty = function (property) {
            return Object.prototype.hasOwnProperty.call(action, property);
        };
        if (hasProperty(service)) {
            if (typeof action[service] === "object") {
                var data_1 = action.target + "Data";
                var error_1 = action.target + "Error";
                var loading = action.target + "Loading";
                var initialData = hasProperty(initialState)
                    ? action[initialState]
                    : false;
                var payload_1 = (_a = {},
                    _a[data_1] = initialData,
                    _a[error_1] = false,
                    _a[loading] = false,
                    _a);
                next(__assign(__assign({}, action), { payload: __assign(__assign({}, payload_1), (_b = {}, _b[loading] = true, _b)) }));
                action[service]
                    .then(function (resp) {
                    var _a;
                    if (resp.status < 200 || resp.status > 299) {
                        var error_2 = new Error();
                        error_2.error = resp;
                        throw error_2;
                    }
                    next(__assign(__assign({}, action), { payload: __assign(__assign({}, payload_1), (_a = {}, _a[data_1] = action.response ? action[response](resp) : resp, _a)) }));
                    if (hasProperty(success))
                        action[success](resp);
                })
                    .catch(function (err) {
                    var _a;
                    next(__assign(__assign({}, action), { payload: __assign(__assign({}, payload_1), (_a = {}, _a[error_1] = action.error ? action.error(err) : err, _a)) }));
                    if (hasProperty(failure))
                        action[failure](err);
                });
            }
            else {
                throw new Error("action.service should be a promise");
            }
        }
        else {
            next(action);
        }
    }; }; };
}
exports.default = responseMiddleware;
//# sourceMappingURL=index.js.map